### PR TITLE
fix(observable): inner property should be non-enumerable

### PR DESCRIPTION
**Pull request by Gheoan**
Fix https://github.com/aurelia/binding/issues/505.

### CODE REVIEW

**Comment by Gheoan** [comment]
Why is the inner property on the prototype instead of the instance?

**Comment by jdanyow** [comment]
@Gheoan nice! This looks good to me.

@jods4, thoughts?

**Comment by jods4** [comment]
@Gheoan because `target` is the prototype of the class for non-static members, that is expect.

I have not run the code, but I think it doesn't work. 
Please add a test that checks if the backing field is still non-enumerable _after_ its value has been changed by client code.

File: test/decorator-observable.spec.js [hunk 1]
```diff
@@ -108,15 +112,41 @@ describe('observable decorator', () => {
...
-      keys.push(p);
-    }
-    expect(keys).toContain('value');
+
+    expect(Object.getPrototypeOf(instance).propertyIsEnumerable('value')).toBe(true);
+  });
+
+  describe('inner property', () => {
+    it('should exist', () => {
+      const instance = new class {
```
Comment (atsu85) [inline test/decorator-observable.spec.js]: Why is the const called `instance` when controversially it seems to be a class, not instance of the class?

**Comment by doktordirk** [comment]
has that been tried with babel?

File: test/decorator-observable.spec.js (hunk 1 above)
Comment (Gheoan) [inline test/decorator-observable.spec.js]: It is an instance. Note the `new` keyword.

Comment (jdanyow) [inline test/decorator-observable.spec.js]: right- anonymous class

**Comment by jdanyow** [comment]
@Gheoan can you add the test @jods4 requested, assuming it passes we can merge.

Thank you both for all of your work on this!

File: test/decorator-observable.spec.js (hunk 1 above)
Comment (Gheoan) [inline test/decorator-observable.spec.js]: It is an instance. Note the `new` keyword.

**Comment by Gheoan** [comment]
Added more tests. @jods4 let me know if this looks good.

**Comment by jods4** [comment]
@Gheoan I am not sure the `Object.propertyIsEnumerable()` test is good enough. It does always return `false` for properties inherited from prototypes. 

When I was prototyping what you've done, I was worried because of this code:
```javascript
function X() {}
Object.defineProperty(X.prototype, 'y', { configurable: true, enumerable: false, writable: true });
var x = new X();
x.y = 42;
Object.propertyIsEnumerable(x, 'y'); // -> false
for (k in x) console.log(x); // -> 'y'
JSON.stringify(x);  // -> { "y": 42 }
```

**Comment by Gheoan** [comment]
@jods4 In the example from your [comment](https://github.com/aurelia/binding/pull/506#issuecomment-250417982): 
- on line 2 a **non-enumerable** property is created on the prototype of `Function` object `X` (capital).
- on line 4 a property is created on the object `x`. Properties created this way are enumerable by default (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Enumerability_and_ownership_of_properties).
- line 5 is not quite right. The rigth way would be `x.propertyIsEnumerable('y')` which is `true` because of line 2.

I will check to be sure that `SetterObserver` does create the accesors on the right object to prevent what happens on line 2 of your example.

**Comment by Gheoan** [comment]
Changed tests to reflect my comment above.

File: test/decorator-observable.spec.js [hunk 2]
```diff
@@ -108,15 +111,69 @@ describe('observable decorator', () => {
...
+        '_value'
+      );
+
+      it('should convert', () => {
+        spyOn(Logger.prototype, 'warn');
+        observer.convertProperty();
+        expect(Logger.prototype.warn).not.toHaveBeenCalled();
+      });
+
+      it('should exist', () => {
```
Comment (Gheoan) [inline test/decorator-observable.spec.js]: This test asserts that `SetterObserver` creates the getter-setter on the right object (https://github.com/aurelia/binding/pull/506#issuecomment-250547363).

**Comment by jods4** [comment]
@Gheoan but isn't it exactly what happens in the decorator code?
  ```ts
Reflect.defineProperty(target, innerPropertyName, {
  configurable: true,
  enumerable: false,
  writable: true
});
```
Here `target` is the prototype of the class, right?
```ts
descriptor.set = function(newValue) {
  let oldValue = this[innerPropertyName];
  this[innerPropertyName] = newValue;
  if (this[callbackName]) {
    this[callbackName](newValue, oldValue, key);
  }
};
```
And here, the first time the value is set, `this[innerPropertyName]` will create an enumerable property directly on the instance?

**Comment by Gheoan** [comment]
Right, I changed the setter to use `target` instead of `this` and added a test.

**Comment by jdanyow** [comment]
@Gheoan `target` cannot be used can it? Wouldn't that make all instances of the class have the same value because target is the prototype?

File: src/decorator-observable.js [hunk 3]
```diff
@@ -36,10 +42,10 @@ export function observable(targetOrConfig: any, key: string, descriptor?: Proper
     delete descriptor.initializer;
 
     // add the getter and setter to the property descriptor.
-    descriptor.get = function() { return this[innerPropertyName]; };
+    descriptor.get = function() { return target[innerPropertyName]; };
     descriptor.set = function(newValue) {
-      let oldValue = this[innerPropertyName];
-      this[innerPropertyName] = newValue;
+      let oldValue = target[innerPropertyName];
+      target[innerPropertyName] = newValue;
```
Comment (jdanyow) [inline src/decorator-observable.js]: @Gheoan I don't think you can use `target` in this way. It creates a closure and is assigning the value to the prototype. I think if you write a test where you have two instances of an `@observable` decorated class, and you set the property on one instance, the second instance will have the same property value.

**Comment by Gheoan** [comment]
Did some changes. The private property is added first to the prototype, then to the instance. It is nonenumerable in both cases.

**Comment by jods4** [comment]
@Gheoan Clever! JS is so tricky sometimes... so you set the property value normally so it works on both direct access and getter/setters, and then set _only_ the `enumerable` flag to false to hide it. I guess that works in all cases, good idea.

File: src/decorator-observable.js [hunk 4]
```diff
@@ -9,6 +9,11 @@ export function observable(targetOrConfig: any, key: string, descriptor?: Proper
 
     // use a convention to compute the inner property name
     let innerPropertyName = `_${key}`;
+    const innerPropertyDescriptor: PropertyDescriptor = {
+      configurable: true,
+      enumerable: false,
+      writable: true,
+    };
```
Comment (jods4) [inline src/decorator-observable.js]: Is this still useful with the new implementation of the setter? What happens if you remove it?

Comment (Gheoan) [inline src/decorator-observable.js]: Do you mean not creating the inner property? Doing so causes some test that were not added in this PR to fail.

Comment (jods4) [inline src/decorator-observable.js]: OK. Maybe they should be included or documented then? It's not obvious why this is still required (I could probably come up with at least one funny case involving `with` binding or overriden contextes).

Comment (Gheoan) [inline src/decorator-observable.js]: Sorry for my confusing comment above. Removing [this](https://github.com/aurelia/binding/pull/506/files#diff-26d3f601c99bc9cd08a30349643a0ac2R44) line causes some old, already existing test to fail.

Comment (Gheoan) [inline src/decorator-observable.js]: Sorry for my confusing comment above. Removing [this](https://github.com/aurelia/binding/pull/506/files#diff-26d3f601c99bc9cd08a30349643a0ac2R44) line causes some old, already existing test to fail.

**Comment by jdanyow** [comment]
@Gheoan thanks for all the hard work 👍

**Pull request merged**